scipy/stats/_qmvnt.py

```diff
@@ -0,0 +1,506 @@
+    """
+    cho, lo, hi = _permuted_cholesky(covar, low, high)
+    n = cho.shape[0]
+    ndim_integrand = n - 1
+    ct = cho[0, 0]
+    c = phi(lo[0] / ct)
+    d = phi(hi[0] / ct)
+    ci = c
+    dci = d - ci
+
+    def integrand(*zs):
+        ndim_qmc = len(zs)
+        n_qmc_samples = len(np.atleast_1d(zs[0]))
+        assert ndim_qmc == ndim_integrand
+        y = np.zeros((ndim_qmc, n_qmc_samples))
+        c = np.full(n_qmc_samples, ci)
+        dc = np.full(n_qmc_samples, dci)
+        pv = dc.copy()
+        for i in range(1, n):
+            if use_tent:
+                # Tent periodization transform.
+                x = abs(2 * zs[i-1] - 1)
+            else:
+                x = zs[i-1]
+            y[i - 1, :] = phinv(c + x * dc)
+            s = cho[i, :i] @ y[:i, :]
+            ct = cho[i, i]
+            c = phi((lo[i] - s) / ct)
+            d = phi((hi[i] - s) / ct)
+            dc = d - c
+            pv = pv * dc
+        return pv
+
+    return integrand, ndim_integrand
+
+
+def _qmvt(m, nu, covar, low, high, rng, lattice='cbc', n_batches=10):
+    """Multivariate t integration over box bounds.
+
+    Parameters
+    ----------
+    m : int > n_batches
+        The number of points to sample. This number will be divided into
+        `n_batches` batches that apply random offsets of the sampling lattice
+        for each batch in order to estimate the error.
+    nu : float >= 0
+        The shape parameter of the multivariate t distribution.
+    covar : (n, n) float array
+        Possibly singular, positive semidefinite symmetric covariance matrix.
+    low, high : (n,) float array
+        The low and high integration bounds.
+    rng : Generator, optional
+        default_rng(), yada, yada
+    lattice : 'cbc' or 'richtmyer' or callable
+        The type of lattice rule to use to construct the integration points.
+    n_batches : int > 0, optional
+        The number of QMC batches to apply.
+
+    Returns
+    -------
+    prob : float
+        The estimated probability mass within the bounds.
+    est_error : float
+        3 times the standard error of the batch estimates.
+    n_samples : int
+        The number of samples actually used.
+    """
+    sn = max(1.0, np.sqrt(nu))
+    low = np.asarray(low, dtype=np.float64)
+    high = np.asarray(high, dtype=np.float64)
+    cho, lo, hi = _permuted_cholesky(covar, low / sn, high / sn)
+    n = cho.shape[0]
+    prob = 0.0
+    error_var = 0.0
+    q, n_qmc_samples = _get_lattice(lattice, n, max(m // n_batches, 1))
+    i_samples = np.arange(n_qmc_samples) + 1
+    for j in range(n_batches):
+        pv = np.ones(n_qmc_samples)
+        s = np.zeros((n, n_qmc_samples))
+        for i in range(n):
+            # Pseudorandomly-shifted lattice coordinate.
+            z = q[i] * i_samples + rng.random()
+            # Fast remainder(z, 1.0)
+            z -= z.astype(int)
+            # Tent periodization transform.
+            x = abs(2 * z - 1)
```

Comment:
I am really confused by this thing. Seems like the intent is to have a sample which is either quite regular with CBC or Richtmyer. Why not just using a QMC sample? This is an integration problem right?

Reply:
As I said earlier, CBC nearly always seems to give more accurate results faster than Sobol'.

Reply:
Even as the dimensionality increases? I don't understand why, the integrand should be smooth here no? Which would be the perfect scenario for (R)QMC.

Did you use `2**n` points for Sobol'? Did you try Halton or LHS?

Reply:
You can play with the notebook in [the gist I linked above](https://gist.github.com/rkern/8518cb3e3f57fa81d21dbd9eabcfaa87) to investigate on your own.

Yes, even as the dimensionality increases; Sobol' approaches CBC, but never really beats it (except for the occasional random fluctuation).

Yes, I used perfect `2**n` points for Sobol'.

Yes, I tried Halton and LHS. Both are significantly less accurate compared to Sobol'.

By speed, I am referring to the wall-clock time for computing the CDF (so generating the QMC points and evaluating the integral). Since the accuracy of the result is never better than CBC for the same number of points (modulo the rounding to the nearest prime when necessary), it always takes at least as many evaluations, so also "slower" in the convergence sense.

Reply:
The integrand also has at least one singularity on the boundary, which the CBC lattice always avoids, but Sobol' sometimes doesn't, depending on the scrambling.

Reply:
Also, not for nothing, the methods implemented here _are_ RQMC, as [well-explained by our own documentation introducing QMC](http://scipy.github.io/devdocs/reference/stats.qmc.html#introduction-to-quasi-monte-carlo).

Reply:
> Yes, even as the dimensionality increases; Sobol' approaches CBC, but never really beats it (except for the occasional random fluctuation).
> ...
> The integrand also has at least one singularity on the boundary, which the CBC lattice always avoids, but Sobol' sometimes doesn't, depending on the scrambling.

If there are known points to avoid, we can and should work around them. Otherwise it's not a fair comparison.

I will try to run your notebook.

> Also, not for nothing, the methods implemented here are RQMC, as well-explained by our own documentation introducing QMC.

I know, I wrote that with Art 😉 Sobol' is just supposed to converge faster on well behaved integrand.

Not sure what you mean by, not for nothing. As it is coded at the moment it's not really following our framework.

Reply:
> If there are known points to avoid, we can and should work around them. Otherwise it's not a fair comparison.

I guess so. I don't see how that's going to improve the performance of Sobol' in this comparison, though. Whatever you do to avoid the `inf` will likely be at a performance cost for no real improvement in the accuracy over CBC.

I guess you could try to derive a different transformation of the integrand that doesn't have the same boundary issue, but the goal here isn't to make "fair" comparisons between CBC lattices and Sobol' generators overall. It's to implement multivariate-t CDFs.

> Not sure what you mean by, not for nothing. As it is coded at the moment it's not really following our framework.

I wanted to be clear to everyone else about what you were claiming. The way that you wrote "Why not just using a QMC sample?" and "Which would be the perfect scenario for (R)QMC." it appeared that you were claiming that what was implemented was not RQMC at all because it was a lattice rule rather than one of the other methods. I wanted it to be clear to everyone that yes, these are RQMC algorithms. Ones that are well-validated for this particular problem by [the guy who literally wrote the book on the subject](https://link.springer.com/book/10.1007/978-3-642-01689-9). If you want to ask about why I'm not using "existing `QMCEngine` implementations" or "`scipy.stats.qmc` framework", that would be a better phrasing. What's currently implemented is not the whole landscape of reasonable options in the domain.

In my experiments, the existing `QMCEngine`s can only come close to the CBC implementation for this integrand, but do not exceed it as an algorithmic matter (i.e. numerical accuracy and sample efficiency). When we get to the Pythranization or other implementation optimizations, the lattice implementation will likely only get better in comparison due to implementation efficiencies because point generation is inlined in the loop very efficiently. Even if we added a `CBCLattice` `QMCEngine`, I probably wouldn't want to use `qmc_quad()` with it in any case (though I would make sure that I could reuse its lattice generator-finding implementation).

Reply:
One can reintroduce some guarding logic from the original MATLAB code in order to restore good behavior on the boundary. It doesn't change anything about the performance of `qmc_quad()` with any of the `QMCEngine`s, though, just guards against the internal `NaN`s from leaking.

Replace

```python
            c = phi((lo[i] - s) / ct)
            d = phi((hi[i] - s) / ct)
```

with 

```python
            c = np.ones(n_qmc_samples)
            d = np.ones(n_qmc_samples)
            lois = lo[i] - s
            hiis = hi[i] - s
            c[lois < -9*ct] = 0.0
            d[hiis < -9*ct] = 0.0
            lo_mask = abs(lois) < 9*ct
            hi_mask = abs(hiis) < 9*ct
            c[lo_mask] = phi(lois[lo_mask] / ct)
            d[hi_mask] = phi(hiis[hi_mask] / ct)
```

in the various loops.

Reply:
What is `ct`?

When the covariance is singular, I've found that it's not uncommon for `d` below to be zero.  This leads to `y` being `-inf`. Is this the singularity mentioned above?

Without thinking about what is really going on, I tried fixing it by calculating `dc` with `log_ndtr` and `logsumexp`, but then that can underflow. I think the whole calculation can be moved to logspace with the added help of `ndtri_exp`, but it sounds like there may be another option?

Reply:
It has nothing to do with the PDF being degenerate. The particular transformation of the PDF to the unit hypercube just has a cusp at the unit boundaries in some circumstances (particularly when the regular-space bounds are semi-infinite) for general covariance matrices. The naive evaluation yields `lo[i] = -inf` and `s = -inf` creating a `NaN` when they are subtracted. The MATLAB code's extra masking stops the `NaN` from leaking out and replacing it with a 0, but the actual value that comes out is not quite correct for the cusp (which does seem to have a finite, nonzero limiting value).

Reply:
> the goal here isn't to make "fair" comparisons between CBC lattices and Sobol' generators overall. It's to implement multivariate-t CDFs.

A nuance here, the goal is to implement multivariate-t CDF in a maintainable way that maintainers who will need to fix bug can understand and fix.

Also note that the usual "social/progamming" contract we tend to have-at the moment at least-is that when claims are made, it's not up to the reviewer to do the work of proving them. We are going in much greater length on some topics to verify and prove things. And sometimes it even involves authors of papers. Art even wrote a paper after the QMC PR. And we could write one as well with what was done for Boschloo/Barnard. I am not saying we should write a paper for every PR OC and there is a limit to what is reasonable to be asked.

Here you are proposing 2 other QMC methods instead of the one we already have. I think it's legitimate to ask why we should add other methods just for this purpose and not follow our existing infrastructure.

If in this case you are certain it's faster and converge faster. Then ok we have our answer and can accept the additional maintenance cost knowingly.

> When we get to the Pythranization or other implementation optimizations, the lattice implementation will likely only get better in comparison due to implementation efficiencies because point generation is inlined in the loop very efficiently.

I am not sure if pure speed is very relevant here, but ok.

Just a note that our QMC engines are not that slow as being implemented in Cython (https://gist.github.com/tupui/54a57f474a069f685ba85c9036f6b121)

> Even if we added a CBCLattice QMCEngine, I probably wouldn't want to use qmc_quad() with it in any case (though I would make sure that I could reuse its lattice generator-finding implementation).

As said above, fine by me if you maintain this code.

I am personally more familiar with the use of the engines and `qmc_quad`. I don't get the interlacing between samples generation, integration and the rest, but that's fine.